plugin/xmark.vim

```diff
@@ -73,6 +73,20 @@ function! s:grp(buf)
   return '_xmark_buffer_' . a:buf . '_'
 endfunction
 
+function! s:urlencode_char(input)
+  let n = char2nr(a:input)
+  let r = ''
+  while n
+    let r = '0123456789ABCDEF'[n % 16] . r
+    let n = n / 16
+  endwhile
+  return '%'. r
+endfunction
+
+function! s:urlencode(input)
+  return substitute(a:input, '\([^a-zA-Z0-9_.-/]\)', '\=s:urlencode_char(submatch(1))', 'g')
+endfunction
+
```

Comment:
The code can be simplified with the use of `printf`, and you don't need the capturing group in the regex. Also, `.-/` at the end of the pattern should be adjusted since it's interpreted as a character range that does not include `-`.

```vim
substitute(a:input, '[^a-zA-Z0-9_./-]', '\=printf("%%%02X", char2nr(submatch(0)))', 'g')
```